awscli/customizations/s3/s3handler.py

```diff
@@ -367,169 +381,128 @@ def _enqueue_upload_end_task(self, filename, upload_context):
-        elif filename.operation_name == 'download':
-            num_uploads = self._enqueue_range_download_tasks(filename)
-        return num_uploads
+        :type manager: s3transfer.manager.TransferManager
+        :param manager: The transfer manager to use for the upload.
 
-    def _enqueue_range_download_tasks(self, filename, remove_remote_file=False):
+        :type bucket: str
+        :param bucket: The bucket to upload the stream to.
 
-        # Create the context for the multipart download.
-        num_downloads = int(filename.size / self.chunksize)
-        context = tasks.MultipartDownloadContext(num_downloads)
+        :type key: str
+        :param key: The name of the key to upload the stream to.
 
-        # No file is needed for downloading a stream.  So just announce
-        # that it has been made since it is required for the context to
-        # begin downloading.
-        context.announce_file_created()
+        :return: A CommandResult representing the upload status.
+        """
+        expected_size = self.params.get('expected_size', None)
+        subscribers = None
+        if expected_size is not None:
+            # `expected_size` comes in as a string
+            expected_size = int(expected_size)
 
-        # Submit download part tasks to the executor.
-        self._do_enqueue_range_download_tasks(
-            filename=filename, chunksize=self.chunksize,
-            num_downloads=num_downloads, context=context,
-            remove_remote_file=remove_remote_file
-        )
-        return num_downloads
+            # set the size of the transfer if we know it ahead of time.
+            subscribers = [ProvideSizeSubscriber(expected_size)]
 
-    def _enqueue_multipart_upload_tasks(self, filename, payload=None):
-        # First we need to create a CreateMultipartUpload task,
-        # then create UploadTask objects for each of the parts.
-        # And finally enqueue a CompleteMultipartUploadTask.
-        if self.params['expected_size']:
+            # TODO: remove when this happens in s3transfer
             # If we have the expected size, we can calculate an appropriate
             # chunksize based on max parts and chunksize limits
-            chunksize = find_chunksize(int(self.params['expected_size']),
-                                       self.chunksize)
+            chunksize = find_chunksize(
+                expected_size, self.config.multipart_chunksize)
         else:
+            # TODO: remove when this happens in s3transfer
             # Otherwise, we can still adjust for chunksize limits
-            chunksize = adjust_chunksize_to_upload_limits(self.chunksize)
+            chunksize = adjust_chunksize_to_upload_limits(
+                self.config.multipart_chunksize)
+        self.config.multipart_chunksize = chunksize
 
-        num_uploads = '...'
+        stream_filein = sys.stdin
+        if six.PY3:
```

Comment:
Yeah there is not really any overlap with ``BinaryStdout``. I think the approach you went with is appropriate.